src/mds/Server.cc

```diff
@@ -2139,7 +2197,53 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
 
   if (mdr->killed) {
     dout(10) << "request " << *mdr << " was killed" << dendl;
-    return;
+    //if the mdr is a "batch_op" and it has followers, pick a follower as
+    //the new "batch_head" and go on processing the new one.
+    if (mdr->is_batch_op() && mdr->is_batch_head && mdr->batch_reqs.size()) {
+      MDRequestRef new_batch_head;
+      for (auto req : mdr->batch_reqs) {
+	mdr->batch_reqs.erase(req);
+	if (!req->killed) {
+	  new_batch_head = req;
+	  break;
+	}
+      }
+
+      if (!new_batch_head) 
+	return;
```

Comment:
all requests are killed. need to cleanup them

Reply:
@ukernel Hi, thanks for your review. I thought those requests should either have already been cleaned up when they are marked as killed, as they can't be "committing". Am I right about this?

Reply:
request_kill() just mark it.  killed request gets cleaned up when it's being dispatched.

Reply:
@ukernel Hi, I don't quite understand, according the code, shown as follows, request_kill will clean up the request as long as it's not commiting.
```
void MDCache::request_kill(MDRequestRef& mdr) {
	// rollback slave requests is tricky. just let the request proceed.
	if (mdr->done_locking && mdr->has_more()
			&& (!mdr->more()->witnessed.empty()
					|| !mdr->more()->waiting_on_slave.empty())) {
		dout(10)
					<< "request_kill " << *mdr
							<< " -- already started slave requests, no-op"
							<< dendl;

		assert(mdr->used_prealloc_ino == 0);
		assert(mdr->prealloc_inos.empty());

		mdr->session = NULL;
		mdr->item_session_request.remove_myself();
		return;
	}

	mdr->killed = true;
	mdr->mark_event("killing request");

	if (mdr->committing) {
		dout(10)
					<< "request_kill " << *mdr
							<< " -- already committing, no-op" << dendl;
	} else {
		dout(10)
					<< "request_kill " << *mdr << dendl;
		request_cleanup(mdr);
	}
}
```

Reply:
you are right, sorry.

Reply:
found an issue here

if new_batch_head is null. we should call  mdr->in[0]->batch_ops.erase(mask) or mdr->dn[0].back()->batch_ops.erase(mask);

Reply:
@ukernel Yeah, I'll correct that:-)